consensus/src/consensus_observer/observer/consensus_observer.rs

```diff
@@ -548,12 +599,37 @@ impl ConsensusObserver {
             return;
         };
 
+        // Get the epoch and round of the first block
+        let first_block = ordered_block.first_block();
+        let first_block_epoch_round = (first_block.epoch(), first_block.round());
+
+        // Determine if the block is behind the last ordered block, or if it is already pending
+        let last_ordered_block = self.get_last_ordered_block();
+        let block_out_of_date =
+            first_block_epoch_round <= (last_ordered_block.epoch(), last_ordered_block.round());
+        let block_pending = self
+            .pending_block_store
+            .lock()
+            .existing_pending_block(&ordered_block);
+
+        // If the block is out of date or already pending, ignore it
+        if block_out_of_date || block_pending {
```

Comment:
not sure how expensive the verify is, but we can move this filter above the verify?

Reply:
I have it this way because `verify_ordered_blocks()` makes sure there is at least one ordered block in the message (we need to verify this first, otherwise we'll panic on `first_block()`). 

Thankfully `verify_ordered_blocks()` is pretty cheap (it doesn't do much other than verify a couple simple properties). I could add a dedicated check that there is at least one ordered block in the message, and move the other verifications later on, but it might be overkill given how cheap `verify_ordered_blocks()` is anyway 🤔